enhancements/bootimages.md

```diff
@@ -0,0 +1,163 @@
+
+OpenShift 4 comes as a "release image" which is a *container* image - bootimages do not naturally fit into that, and currently the installer has some ad-hoc support for dealing with bootimages.
+
+A goal here is to make bootimage management more of a first class operation, something like:
+
+```
+$ oc adm release generate-bootimage quay.io/openshift/release:4.3 vmware
+```
+
+Rather than having an administrator [manually download a bootimage](https://mirror.openshift.com/pub/openshift-v4/dependencies/rhcos/4.3/latest/) corresponding to the release version, this command would output a bootimage for the chosen platform/media.
+
+Similarly:
+
+```
+$ oc adm release generate-bootimage quay.io/openshift/release:4.3 aws
+```
+
+would output data similar to the [aws rhcos.json data](https://github.com/openshift/installer/blob/2055609f95b19322ee6cfdd0bea73399297c4a3e/data/data/rhcos.json#L2) so that AWS UPI installations could use it programatically.
+
+Other output proposals
+
+- `metal`, `openstack`, `qemu`: What's available today for 4.3
+- `aws-snapshot` to handle [private AWS regions](https://github.com/openshift/os/blob/411d1f5943ea23f2de7e4f7a04ab0bb185fd9586/FAQ.md#q-how-do-i-get-rhcos-in-a-private-ec2-region).
+- `baremetal-iso` generates a "Live" ISO as is used in Fedora CoreOS and points people at how to do [Ignition injection](https://github.com/coreos/coreos-installer/blob/3652e6a767bad593b1b898f239e41bc11a83ab8f/src/iso.rs#L28)
+- `baremetal-pxe` outputs the split kernel/initramfs suitable for PXE
+
+### Non-Goals
+
+#### Replacing the default in-place update path
+
+In-place updates as [managed by the MCO](https://github.com/openshift/machine-config-operator/blob/master/docs/OSUpgrades.md) today works fairly seamlessly.  We can't require that everyone fully reprovision a machine in order to do in-place updates - that makes updates *much* more expensive.  It implies re-downloading all container images, etc.
+
+#### Exposing a general purpose "OS build" tool
+
+This proposal includes shipping a subset of https://github.com/coreos/coreos-assembler/ - but that's not the same as supporting custom "CoreOS-style" builds for non-OpenShift use cases.
+
+## Proposal phase 1: bootimage.json in release image
+
+First, the RHCOS build process is changed to inject the current coreos-assembler `meta.json` output for the build into `machine-os-content`.  This aims to move the "source of truth" for cluster bootimages into the release image.  Nothing would use this to start - however, as soon as we switch to a machine API provisioned control plane, having that process consume this data would be a natural next step.
+
+In fact, we could aim to switch to having workers use the `bootimage.json` from the release payload immediately after it lands.  A downside is this would open up potential for drift between the bootstrap+controlplane and workers.
+
+## Proposal phase 2: oc adm release generate-bootimage
```

Comment:
Can we include details on a option that prints the contents of bootimage.json like mentioned here https://github.com/openshift/enhancements/pull/201/files#diff-9b570624a774881c701bbfbc18b60109R64

a way to get the prebuilt bootimages if the user sees fit as an alternative to generating the whole thing...?

Reply:
This gets into a fundamental question of whether we do ship pre-built bootimages going forward in general.  Clearly in "phase 1" we continue to do so as we can't break the way openshift-install works today.

The thing is though in any disconnected environment it'd require another step for admins to mirror on an ongoing basis.  

On one hand though, we absolutely are going to need to make it ergonomic at some point to "boot a single RHCOS node" for testing.  There are many valid reasons for this, among them hardware validation before trying a cluster install, experimenting with configuration "live" that one then compiled into Ignition/MachineConfig, etc.  And until such time as we have `openshift-install run-one-coreos-instance` or something...people are going to want to download the bootimages directly and boot them too.

So...I'm conflicted.

Maybe what happens in the end is we do do both - but the cadence of "golden bootimages" would still be once per release (for the bootstrap).  There's no point to us uploading e.g. new AMIs if in practice clusters are going to use in-place update mechanisms anyways.

IOW...I think people doing UPI installs would turn to automating `oc adm release generate-bootimage`.

Reply:
For phase 1, i think it's necessary for migration.

we could add a `--info` flag with it defaulting to true to begin with. and when we have actual generation capability we move to `--info` false so users are generating the images.

i think `--info` will continue to provide value in future, because it will allow the customers to get information like

what's the rhcos version, maybe what's included, if there are any publicly available images they are try let's use them. and as long as the output is versioned users have a way to differentiate and automate.

Reply:
OK I'm thinking about this again.  I guess one thing here is what you *most* strongly objected to was having `openshift-install` be the frontend for this information (clearly we want to do all the other stuff here though too).

But clearly a "MVP" for this is basically what you're saying with `--info`.  The question is how to get there from the today where RHCOS is pinned in the installer.  I really don't want two sources of this information (it's already bad enough docs are updated manually).  So...what if we still added a hidden/secret mechanism for `oc adm release new` to download `openshift-install` and extract this data from it somehow?  

The simplest would perhaps be a hidden version of `openshift-install hidden-only-for-oc-output-coreos-build` or so.

Or, we could stick it in a special ELF section (would require `oc adm` to vendor some ELF tools).

Or, we could do a variant of https://github.com/openshift/installer/pull/1422 and basically scrape the binary; the lowest tech approach...

Reply:
I think we should include an object into the release image like we include the machine-os-content.

This object can then be used to provide the information as part of `oc adm release info --boot-images`

trying to get this data from installer is not correct imo. the installer's embedded data is mostly for bootstrap-host and it currently being used for cluster hosts.

Reply:
Yes, we are all agreed on the long term goal of getting the data out of the installer.  The problem is in the *short* term, having two sources of truth for this data is going to be very problematic.

Now...hmm, we could get the data into a container in the release payload and change `oc adm release new` to patch the `openshift-install` binary, the same way it does for the release image itself?

Reply:
The downside of doing this is that anyone building `openshift-install` from git...well, it just won't work.  I guess we could change the installer to fetch "the latest" from a URL if built from git, basically the same way we do with the release image.

Reply:
For metal IPI CI, we stuff the data into the baremetal-installer container because **we have to** have it. See: https://github.com/openshift/installer/pull/3330

The way users get this data to use metal IPI is really painful (`openshift-install version`, and fetching the rhcos.json from GitHub for that sha!).  Basically every on-premise platform needs access to the image locations...

I dislike client-side patching of the installer binary. Storing it somehow next to machine-os-content makes sense to me.

Reply:
> I dislike client-side patching of the installer binary.

I'm not talking about any client side patching.  This is something that would happen in the OpenShift build system (release controller).

> Storing it somehow next to machine-os-content makes sense to me.

Yes but the problem with that is that it's the bootstrap node that pulls the release image, so that leaves open the problem of which image to use for the bootstrap.

Now I guess a first transitional step could be changing the installer to use the metadata from the release payload for the cluster, and leave the pinned version only for the bootstrap node.

Reply:
>> I dislike client-side patching of the installer binary.
> I'm not talking about any client side patching. This is something that would happen in the OpenShift build system (release controller).

`oc adm release extract` patches the binary. Sure, the binaries published on mirror.openshift.com already had that run to get them, but users can do that themselves too clientside.

It's the default for baremetal since we don't publish ours.

>>  Storing it somehow next to machine-os-content makes sense to me.
> Yes but the problem with that is that it's the bootstrap node that pulls the release image, so that leaves open the problem of which image to use for the bootstrap.

A user can just use `oc adm release info`? That's at least a consistent place for any release-related information.